// browser-vite/dist/node/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var build = require('./chunks/dep-7afecafa.js');
require('fs');
require('path');
require('os');
require('tty');
require('util');
require('net');
require('events');
require('url');
require('http');
require('stream');
require('resolve');
require('module');
require('perf_hooks');
require('https');
require('zlib');
require('crypto');
require('tls');
require('assert');
require('buffer');
require('querystring');
require('esbuild');
require('child_process');
require('worker_threads');
require('readline');



exports.build = build.build;
exports.createDevHtmlTransformFn = build.createDevHtmlTransformFn;
exports.createLogger = build.createLogger;
exports.createPluginContainer = build.createPluginContainer;
exports.createServer = build.createServer;
exports.defineConfig = build.defineConfig;
exports.flattenId = build.flattenId;
exports.generateCodeFrame = build.generateCodeFrame;
exports.injectQuery = build.injectQuery;
exports.loadConfigFromFile = build.loadConfigFromFile;
exports.loadEnv = build.loadEnv;
exports.mergeAlias = build.mergeAlias;
exports.mergeConfig = build.mergeConfig;
exports.normalizePath = build.normalizePath;
exports.optimizeDeps = build.optimizeDeps;
exports.posToNumber = build.posToNumber;
exports.preview = build.preview;
exports.printHttpServerUrls = build.printHttpServerUrls;
exports.removeImportQuery = build.removeImportQuery;
exports.resolveBaseUrl = build.resolveBaseUrl;
exports.resolveConfig = build.resolveConfig;
exports.resolveEnvPrefix = build.resolveEnvPrefix;
exports.resolvePackageData = build.resolvePackageData;
exports.resolvePackageEntry = build.resolvePackageEntry;
exports.scanImports = build.scanImports;
exports.searchForWorkspaceRoot = build.searchForWorkspaceRoot;
exports.send = build.send;
exports.sortUserPlugins = build.sortUserPlugins;
exports.transformWithEsbuild = build.transformWithEsbuild;
exports.unwrapId = build.unwrapId;
